server/src/main/java/org/opensearch/common/cache/provider/CacheProvider.java

```diff
@@ -0,0 +1,91 @@
+import org.opensearch.plugins.CachePlugin;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Holds all the cache factories and provides a way to fetch them when needed.
+ */
+public class CacheProvider {
+
+    private final Map<CacheStoreType, List<StoreAwareCache.Factory>> cacheStoreTypeFactories;
+
+    private final Settings settings;
+
+    public CacheProvider(List<CachePlugin> cachePlugins, Settings settings) {
+        this.cacheStoreTypeFactories = getCacheStoreTypeFactories(cachePlugins);
+        this.settings = settings;
+    }
+
+    private Map<CacheStoreType, List<StoreAwareCache.Factory>> getCacheStoreTypeFactories(List<CachePlugin> cachePlugins) {
+        Map<CacheStoreType, List<StoreAwareCache.Factory>> cacheStoreTypeFactories = new HashMap<>();
+        for (CachePlugin cachePlugin : cachePlugins) {
+            Map<CacheStoreType, StoreAwareCache.Factory> factoryMap = cachePlugin.getCacheStoreTypeMap();
+            for (Map.Entry<CacheStoreType, StoreAwareCache.Factory> entry : factoryMap.entrySet()) {
+                cacheStoreTypeFactories.computeIfAbsent(entry.getKey(), k -> new ArrayList<>()).add(entry.getValue());
+            }
+        }
+        return Collections.unmodifiableMap(cacheStoreTypeFactories);
+    }
+
+    public Map<CacheStoreType, List<StoreAwareCache.Factory>> getCacheStoreTypeFactories() {
+        return cacheStoreTypeFactories;
+    }
+
+    /**
+     * Given a map of storeType and cacheName setting, extract a specific implementation.
+     * type.
+     * @param cacheStoreTypeSettings Setting map
+     * @return CacheStoreType
+     */
+    public Map<CacheStoreType, StoreAwareCache.Factory> getCacheStoreType(Map<CacheStoreType, Setting<String>> cacheStoreTypeSettings) {
+        Map<CacheStoreType, StoreAwareCache.Factory> cacheStoreTypeFactoryMap = new HashMap<>();
+        for (Map.Entry<CacheStoreType, List<StoreAwareCache.Factory>> cacheStoreTypeFactoryEntry : cacheStoreTypeFactories.entrySet()) {
+            CacheStoreType cacheStoreType = cacheStoreTypeFactoryEntry.getKey();
+            if (!cacheStoreTypeSettings.containsKey(cacheStoreType)) {
+                continue;
+            }
+            for (StoreAwareCache.Factory factory : cacheStoreTypeFactoryEntry.getValue()) {
+                if (factory.getCacheName().equals(cacheStoreTypeSettings.get(cacheStoreType).get(settings))) {
+                    cacheStoreTypeFactoryMap.put(cacheStoreType, factory);
+                    break;
+                }
+            }
+        }
+        return cacheStoreTypeFactoryMap;
+    }
+
+    /**
+     * Cache types available.
+     */
+    public enum CacheType {
+        EHCACHE("ehcache");
+
+        private final String value;
+
+        CacheType(String value) {
+            this.value = value;
+        }
+
+        public String getValue() {
+            return value;
+        }
+    }
```

Comment:
You can't do this. It defeats the whole purpose of plugins if core needs to know what caches are available.

This provider receives the cache factories from the plugins and registers them in a map. Then, when we create a cache by name, the provider should look for a cache factory for that name (throwing an exception if it doesn't have one), and use that factory to instantiate the cache (possibly passing some configuration to the factory without trying to understand it along the way).

Reply:
Agree on that, certainly looks hacky to me as well.

We can have the provider holding all list of factories as you mentioned.

I was thinking about the actual integration for example in IndicesRequestCache or elsewhere in case it needs to implement a specific version of cache and considering it is not aware of the available cache variants.

We can probably have a setting for IRC which says ```disk.cache.name = ehcache``` which needs to be defined at yml file?

Reply:
Removed this.